packages/core/src/storage/extract.ts

```diff
@@ -17,62 +16,77 @@ export function extractStorageLayout(
   contractDef: ContractDefinition,
   decodeSrc: SrcDecoder,
   deref: ASTDereferencer,
+  storageLayout?: StorageLayout | undefined,
 ): StorageLayout {
   const layout: StorageLayout = { storage: [], types: {}, layoutVersion: currentLayoutVersion };
+  if (storageLayout !== undefined) {
+    layout.types = storageLayout.types;
 
-  // Note: A UserDefinedTypeName can also refer to a ContractDefinition but we won't care about those.
-  const derefUserDefinedType = deref(['StructDefinition', 'EnumDefinition']);
-
-  for (const varDecl of contractDef.nodes) {
-    if (isNodeType('VariableDeclaration', varDecl)) {
-      if (!varDecl.constant && varDecl.mutability !== 'immutable') {
-        const type = normalizeTypeIdentifier(typeDescriptions(varDecl).typeIdentifier);
-
+    for (const storage of storageLayout.storage) {
+      const varDecl = contractDef.nodes.filter(n => n.id == storage.astId && isNodeType('VariableDeclaration', n))[0];
+      if (varDecl) {
         layout.storage.push({
           contract: contractDef.name,
-          label: varDecl.name,
-          type,
+          label: storage.label,
+          type: storage.type,
           src: decodeSrc(varDecl),
         });
```

Comment:
```suggestion
        const { label, offset, slot, type } = storage;
        const src = decodeSrc(varDecl);
        const contract = /* ??? */;
        layout.storage.push({ label, offset, slot, type, contract, src });
```

Note: `contract` must be the contract where the variable is defined, which is not `contractDef.name`, because `storageLayout.storage` is already linearized.

Reply:
Idea: iterate over `contractDef.linearizedBaseParents` and see which one contains this variable declaration.